File: shared/src/commonMain/kotlin/com/aqeel/catsapp_kmm/viewmodel/CatsViewModel.kt
package com.aqeel.catsapp_kmm.viewmodel

import com.aqeel.catsapp_kmm.model.Cat
import com.aqeel.catsapp_kmm.repository.CatsRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch

class CatsViewModel(
    private val repository: CatsRepository,
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Main)
) : CatsViewModelInterface {
    override val cats: Flow<List<Cat>> = repository.getAllCats()

    override fun addCat(name: String, breed: String, age: Int, imageUrl: String?, description: String?) {
        val newCat = Cat(
            name = name,
            breed = breed,
            age = age,
            imageUrl = imageUrl,
            description = description
        )
        scope.launch {
            repository.addCat(newCat)
        }
    }

    override fun deleteCat(id: String) {
        scope.launch {
            repository.deleteCat(id)
        }
    }
} 

File: iosApp/iosApp/CatsViewModel.swift
import shared
import Combine

@MainActor
class CatsViewModelWrapper: ObservableObject {
    @Published private(set) var cats: [Cat] = []
    private let viewModel: CatsViewModel
    
    init(viewModel: CatsViewModel) {
        self.viewModel = viewModel
        observeCats()
    }
    
    private func observeCats() {
        viewModel.cats.watch { [weak self] catsList in
            if let cats = catsList as? [Cat] {
                self?.cats = cats
            }
        }
    }
    
    func addCat(name: String, breed: String, age: Int32) {
        viewModel.addCat(name: name, breed: breed, age: age, imageUrl: nil, description: nil)
    }
    
    func deleteCat(id: String) {
        viewModel.deleteCat(id: id)
    }
} 